ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCuda.java

```diff
@@ -0,0 +1,543 @@
+
+   private FloatPointer groundToWorldTransformHostPointer;
+   private FloatPointer groundToWorldTransformDevicePointer;
+
+   private FloatPointer parametersHostPointer;
+   private FloatPointer parametersDevicePointer;
+
+   private CUDAKernel updateKernel;
+   private CUDAKernel registerKernel;
+   private CUDAKernel croppingKernel;
+   private CUDAKernel preprocessKernel;
+
+   private dim3 blockSize;
+   private dim3 gridSizeKernel0;
+   private dim3 gridSizeKernel1;
+   private dim3 gridSizeKernel2;
+   private dim3 gridSizeKernel3;
+
+   float[] paramsArray;
+
+   private boolean processing = false;
+
+   public void setModeSpherical()
+   {
+      mode = 0;
+   }
+
+   public void setModePerspective()
+   {
+      mode = 1;
+   }
+
+   public RapidHeightMapExtractorCuda(ReferenceFrame leftFootSoleFrame, ReferenceFrame rightFootSoleFrame)
+   {
+      footSoleFrames.put(RobotSide.LEFT, leftFootSoleFrame);
+      footSoleFrames.put(RobotSide.RIGHT, rightFootSoleFrame);
+   }
+
+   public void setDepthIntrinsics(CameraIntrinsics cameraIntrinsics)
+   {
+      this.cameraIntrinsics = cameraIntrinsics;
+   }
+
+   public void recomputeDerivedParameters()
+   {
+      centerIndex = HeightMapTools.computeCenterIndex(heightMapParameters.getLocalWidthInMeters(), heightMapParameters.getLocalCellSizeInMeters());
+      localCellsPerAxis = 2 * centerIndex + 1;
+      gridOffsetX = (float) heightMapParameters.getLocalWidthInMeters() / 2.0f;
+      globalCenterIndex = HeightMapTools.computeCenterIndex(heightMapParameters.getInternalGlobalWidthInMeters(),
+                                                            heightMapParameters.getInternalGlobalCellSizeInMeters());
+      globalCellsPerAxis = 2 * globalCenterIndex + 1;
+
+      cropCenterIndex = (heightMapParameters.getCropWindowSize() - 1) / 2;
+
+      if (2 * cropCenterIndex + 1 != heightMapParameters.getCropWindowSize())
+         throw new RuntimeException("The crop center index was computed incorrectly.");
+   }
+
+   public int getSequenceNumber()
+   {
+      return sequenceNumber;
+   }
+
+   public boolean isInitialized()
+   {
+      return initialized;
+   }
+
+   public TerrainMapData getTerrainMapData()
+   {
+      return terrainMapData;
+   }
+
+   public Point3D getSensorOrigin()
+   {
+      return sensorOrigin;
+   }
+
+   public static HeightMapParameters getHeightMapParameters()
+   {
+      return heightMapParameters;
+   }
+
+   private void deallocateFloatPointer(FloatPointer hostPointer, Pointer devicePointer)
+   {
+      if (hostPointer != null)
+      {
+         hostPointer.close();
+         System.out.println("Deallocated host pointer.");
+      }
+
+      if (devicePointer != null)
+      {
+         devicePointer.close();
+         System.out.println("Deallocated device pointer.");
+      }
+   }
+
+   public void initialize() throws URISyntaxException
+
+   {
+      stream = new CUstream_st();
+      cudart.cudaStreamCreate(stream);
```

Comment:
You should use the `CUDAStreamManager` instead (the 2D CUDA example show how it's used).